### PR TITLE
a style section on naming

### PR DESCRIPTION
**Pull request by mlubin**
@IainNZ @odow @blegat @joaquimg @IssamT @rdeits @tkoolen

### CODE REVIEW

**Comment by mlubin** [comment]
Not currently covered: 

- how to name type parameters (they're local variables but should they be named like types?)
- when it's ok to use 1-letter variable names

File: docs/src/style.md [hunk 1]
```diff
@@ -83,6 +88,55 @@ empty vector with element type `T`. Prefer `T[]` because it is more concise.
...
+The Julia style guide recommends avoiding underscores "when readable", for
+example, `haskey`, `isequal`, `remotecall`, and `remotecall_fetch`. This
+convention creates the potential for unnecessary bikeshedding and also forces
+the user to recall the presence/absence of an underscore, e.g., "was that
+argument named `basename` or `base_name`?". For consistency, *always use
+underscores* in variable names and function names to separate words.
+
+#### Use of `!`
+
+Julia has a convention of appending `!` to a function name if the function
```
Comment (IainNZ) [inline docs/src/style.md]: Couple of days later, still feeling pretty good about this.

File: docs/src/style.md [hunk 2]
```diff
@@ -83,6 +88,55 @@ empty vector with element type `T`. Prefer `T[]` because it is more concise.
...
+`add_constraint` and `set_name`. We allow `!` when it is used to distinguish
+between modifying and non-modifying variants of the same function like
+`scale` and `scale!`. In general, it is more important to clearly document
+which arguments are modified rather than adding `!` to the function name since
+the latter is ambiguous when multiple arguments are present. Never use `!!`.
+
+#### Abbreviations
+
+Abbreviate names to make the code more readable, not to save typing.
+Don't arbitrarily delete letters from a word to abbreviate it (e.g., `indx` or
```
Comment (IainNZ) [inline docs/src/style.md]: Hah but I love abbreviating index to idx...
How do we feel about `for i in 1:10` - is `i` ok? Is `idx` preferable to `i`?
The Google Python style guide IIRC doesn't like `i`, linter complains at least.

Comment (mlubin) [reply docs/src/style.md]: I removed the prohibition on `idx`. Just be consistent. I'll leave discussion on `i` for later.

**Pull request merged**